pkg/runner/cluster_swarm.go

```diff
@@ -1 +1,318 @@
+				Replicas: &replicas,
+			},
+		},
+		Annotations: swarm.Annotations{Name: sname},
+		TaskTemplate: swarm.TaskSpec{
+			ContainerSpec: &swarm.ContainerSpec{
+				Image: image,
+				Env:   env,
+			},
+			RestartPolicy: &swarm.RestartPolicy{
+				Condition: swarm.RestartPolicyConditionNone,
+			},
+			Resources: &swarm.ResourceRequirements{
+				Reservations: &swarm.Resources{
+					MemoryBytes: 60 * 1024 * 1024,
+				},
+				Limits: &swarm.Resources{
+					MemoryBytes: 30 * 1024 * 1024,
+				},
+			},
+			Placement: &swarm.Placement{
+				MaxReplicas: 200,
+			},
+		},
+	}
+
+	logging.S().Infof("fetching an authorization token from AWS ECR")
+
+	// Get an authorization token from AWS ECR.
+	auth, err := aws.ECR.GetAuthToken(input.EnvConfig.AWS)
+	if err != nil {
+		return nil, err
+	}
+
+	logging.S().Infof("fetched an authorization token from AWS ECR")
+
+	scopts := types.ServiceCreateOptions{
+		QueryRegistry: true,
+		// the registry auth will be propagated to all docker swarm nodes so
+		// they can fetch the image properly.
+		EncodedRegistryAuth: aws.ECR.EncodeAuthToken(auth),
+	}
+
+	logging.S().Infow("creating the service on docker swarm", "name", sname, "image", image, "replicas", replicas)
+
+	// Now create the docker swarm service.
+	resp, err := cli.ServiceCreate(ctx, spec, scopts)
+	if err != nil {
+		return nil, err
+	}
+
+	serviceID := resp.ID
+
+	logging.S().Infow("service created successfully", "id", serviceID)
+
+	out := &api.RunOutput{RunnerID: serviceID}
+
+	// If we are running in background mode, return immediately.
+	if cfg.Background {
+		return out, nil
+	}
+
+	// Tail the service until all instances are done, then remove the service if
+	// the flag has been set.
+	rc, err := cli.ServiceLogs(context.Background(), serviceID, types.ContainerLogsOptions{
+		ShowStdout: true,
+		ShowStderr: true,
+		Since:      "2019-01-01T00:00:00",
+		Follow:     true,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed while tailing logs: %w", err)
+	}
+
+	// Docker multiplexes STDOUT and STDERR streams inside the single IO stream
+	// returned by ServiceLogs. We need to use docker functions to separate
+	// those strands, and because we don't care about treating STDOUT and STDERR
+	// separately, we consolidate them into the same io.Writer.
+	rpipe, wpipe := io.Pipe()
+	go func() {
+		_, err := stdcopy.StdCopy(wpipe, wpipe, rc)
+		_ = wpipe.CloseWithError(err)
+	}()
+
+	// This goroutine monitors the state of tasks every two seconds. When all
+	// tasks are shutdown, we are done here. We close the logs io.ReadCloser,
+	// which in turns signals that the runner is now finished.
+	go func() {
+		var errCnt int
+
+		tick := time.NewTicker(2 * time.Second)
+		defer tick.Stop()
+
+		for range tick.C {
+			ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+			tasks, err := cli.TaskList(ctx, types.TaskListOptions{
+				Filters: filters.NewArgs(filters.Arg("service", serviceID)),
+			})
+			cancel()
```

Comment:
someone's going to use this context. We should probably scope `ctx` so nobody does.